library/alloc/src/vec/into_iter.rs

```diff
@@ -158,6 +158,28 @@ impl<T, A: Allocator> Iterator for IntoIter<T, A> {
         (exact, Some(exact))
     }
 
+    #[inline]
+    fn advance_by(&mut self, n: usize) -> Result<(), usize> {
+        let step_size = self.len().min(n);
+        if mem::size_of::<T>() == 0 {
+            // SAFETY: due to unchecked casts of unsigned amounts to signed offsets the wraparound
+            // effectively results in unsigned pointers representing positions 0..usize::MAX,
+            // which is valid for ZSTs.
+            self.ptr = unsafe { arith_offset(self.ptr as *const i8, step_size as isize) as *mut T }
+        } else {
+            let to_drop = ptr::slice_from_raw_parts_mut(self.ptr as *mut T, step_size);
+            // SAFETY: the min() above ensures that step_size is in bounds
+            unsafe {
+                self.ptr = self.ptr.add(step_size);
+                ptr::drop_in_place(to_drop);
+            }
+        }
+        if step_size < n {
+            return Err(step_size);
+        }
```

Comment:
Should this check happen after moving the `ptr`?

Reply:
> Should this check happen after moving the ptr?

Not sure what you mean. It's not really a "check" for correctness, it's about choosing the return value required by the API.

> Wouldn't step_size will always be >= n?

The `min` picks the smaller of the two values, so `step_size <= n`.

Reply:
> Not sure what you mean. It's not really a "check" for correctness, it's about choosing the return value required by the API.

No, I mean do `if step_size < n` earlier, before moving the pointers.

> The min picks the smaller of the two values, so step_size <= n.

Ah, true. I was confused for a bit.

Reply:
> No, I mean do if step_size < n earlier, before moving the pointers.

But wouldn't that prevent the iterator from actually advancing?

Reply:
@pickfire can you explain what issue your suggestion is trying to address?

`advance_by` is supposed to advance as many elements as it can, an `Err(usize)` simply indicates it reached the end of the iterator and hasn't been able to advance over as many elements as requested. Which means it should still try to advance the pointers before returning an error.

Reply:
Ah, I thought it can just return an error without advancing the pointers, is it necessary to advance the pointers if we already know that it will return an error?

Reply:
> Or maybe that's the expected behavior that it advanced the pointers?

Yes, that's what I have been saying. https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.advance_by

Reply:
> is it necessary to advance the pointers if we already know that it will return an error

If you don't advance them then the next time you call `.next()` you'll get the wrong value.